### DISCUSSION TITLE
How Node.js actually responds to request in callbacks?

### DISCUSSION THREAD

**Discussion by MalikHamzaHanif** [nodejs/help#4641] [Q&A]
const http = require("http");
const {readFile}=require("fs")
const path=require("path")
const server = http.createServer((req, res) => {
    if(req.url=="/"){
        readFile(path.resolve(__dirname,"04_table.html"),(err,data)=>{   
               
                 res.write(data)
                 res.end();
                 console.log("callback is completed")
        })
        res.write(`welcome to the homeapge\n`);
       res.end()
    }
});

server.listen(5000, () => {
    console.log("Server is listening");
});

in this code callback is fired and res is sent again after the res.end() is called .why does node is not throwing any error  ? as it is not allowed in nodejs . like you can not sent another res once the res.end() is called. could someone please explain this to me in detail

**Comment by mattpocock** [comment]
>  as it is not allowed in nodejs

Could you clarify what you mean? What behaviour are you seeing here?

**Reply by MalikHamzaHanif** [reply]
Here is case 1
const http = require("http");
const {readFile}=require("fs")
const path=require("path")
const server = http.createServer((req, res) => {
    if(req.url=="/"){
     
        res.write(`welcome to the homepage\n`);
        res.end();
        res.write(`welcome to the homepage\n`);
        res.end();
    }
});

server.listen(5000, () => {
    console.log("Server is listening");
}); 
like when you call res.end() then you can not send any other response to that request if you do so your server will crash like the example mentioned above and as its against http protocols .

but in Case 2 when i am using callback  nodejs is not behaving as expected ..

const http = require("http");
const {readFile}=require("fs")
const path=require("path")
const server = http.createServer((req, res) => {
    if(req.url=="/"){
        readFile(path.resolve(__dirname,"04_table.html"),(err,data)=>{    
                 res.write(data)
                 res.end();
                 console.log("callback is completed")
        })
        res.write(`welcome to the homeapge\n`);
       res.end()
    }
});

server.listen(5000, () => {
    console.log("Server is listening");
});

like i have sent the response and i can see welcome to homepage in my browser and then i have called res.end() .
   res.write(`welcome to the homeapge\n`);
       res.end()
But there is readFile in which i have used a callback which is when fired there is another response i am sending after res.end() but in this particular case the server didnot crashed nor it throws any error .
 readFile(path.resolve(__dirname,"04_table.html"),(err,data)=>{    
                 res.write(data)
                 res.end();
                 console.log("callback is completed")
        })

so why did it doesnot crashed as server has already sent the response .

**Reply by mattpocock** [reply]
Could you reformat this code using proper code blocks (```)? Otherwise it's almost illegible.

**Reply by MalikHamzaHanif** [reply]
Here is case 1

```js
const http = require("http");
const {readFile}=require("fs")
const path=require("path")
const server = http.createServer((req, res) => {
if(req.url=="/"){

    res.write(`welcome to the homepage\n`);
    res.end();
    res.write(`welcome to the homepage\n`);
    res.end();
}
});

server.listen(5000, () => {
console.log("Server is listening");
});
```

like when you call res.end() then you can not send any other response to that request if you do so your server will crash like the example mentioned above and as its against http protocols .

but in Case 2 when i am using callback nodejs is not behaving as expected ..

```js
const http = require("http");
const {readFile}=require("fs")
const path=require("path")
const server = http.createServer((req, res) => {
if(req.url=="/"){
readFile(path.resolve(__dirname,"04_table.html"),(err,data)=>{
res.write(data)
res.end();
console.log("callback is completed")
})
res.write(welcome to the homeapge\n);
res.end()
}
});

server.listen(5000, () => {
console.log("Server is listening");
});
```

like i have sent the response and i can see welcome to homepage in my browser and then i have called res.end() .

```js
res.write(welcome to the homeapge\n);
res.end()
```

But there is readFile in which i have used a callback which is when fired there is another response i am sending after res.end() but in this particular case the server didnot crashed nor it throws any error .

```js
readFile(path.resolve(__dirname,"04_table.html"),(err,data)=>{
res.write(data)
res.end();
console.log("callback is completed")
})
```

so why did it doesnot crashed as server has already sent the response .

**Reply by mattpocock** [reply]
So, I don't know - I'm sure someone could answer. But I'm also wondering what the benefit of knowing this information is? Could I ask why you're asking? Just curiosity?

FWIW, the above code could be better structured using promises instead of callbacks, which you likely know already.

**Reply by MalikHamzaHanif** [reply]
I'm just asking for learning purposes. I want to understand why Node.js behaves this way and to know the underlying reason behind it.
